### PR TITLE
Functionality for non-cash contributions

### PR DESCRIPTION
**Pull request by stuartmccoll**
This pull request adds functionality and validation to the non-cash contributions page of the service journey.

I think there has to be a better way of re-rendering the page and retaining the values when validation has failed rather than using `flash`. I'll raise an issue to revisit this post-MVP.

I've also re-used the code from the error summary partial - it'll need updating to be able to handle nested attributes too. Rather than get bogged down with that here, I'll raise an issue to revisit.

Note that the design isn't in place for the table - one for @paulmsmith to look at separately.

### CODE REVIEW

File: app/controllers/project/project_non_cash_contributions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+class Project::ProjectNonCashContributionsController < ApplicationController
+  include ProjectContext
+
+  def update
+
+    # Empty flash values to ensure that we don't redisplay them unnecessarily
+    flash[:description_value] = ""
```
Comment (gidsg) [inline app/controllers/project/project_non_cash_contributions_controller.rb]: are we sure this is being used, it doesn't seem to be referenced anywhere else?

File: app/views/project/project_non_cash_contributions/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  <div class="nlhf-summary__body">
+
+    <% if @project.non_cash_contributions.first.try(:id).nil? %>
+
+      <h3 class="govuk-heading-m govuk-!-margin-bottom-0">You have not added any non-cash contributions</h3>
+
+    <% else %>
+
+      <% @project.non_cash_contributions.filter { |ncc| ncc.id.present? }.each do |ncc| %>
+
```
Comment (gidsg) [inline app/views/project/project_non_cash_contributions/show.html.erb]: Should we add a TODO here to do a proper design for this section?

File: app/controllers/project/project_non_cash_contributions_controller.rb (hunk 1 above)
Comment (stuartmccoll) [reply app/controllers/project/project_non_cash_contributions_controller.rb]: Good spot! It was supposed to be `flash[:description]` - fixed in the latest commit.

File: app/views/project/project_non_cash_contributions/show.html.erb (hunk 2 above)
Comment (stuartmccoll) [reply app/views/project/project_non_cash_contributions/show.html.erb]: Added 👍.

File: app/controllers/project/project_non_cash_contributions_controller.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@ class Project::ProjectNonCashContributionsController < ApplicationController
   def update
 
     # Empty flash values to ensure that we don't redisplay them unnecessarily
-    flash[:description_value] = ""
+    flash[:description] = ""
```
Comment (gidsg) [inline app/controllers/project/project_non_cash_contributions_controller.rb]: I wonder if using [flash.now](https://guides.rubyonrails.org/action_controller_overview.html#flash-now) would negate the need to instantiate these as empty? But not critical for this PR.

Comment (stuartmccoll) [reply app/controllers/project/project_non_cash_contributions_controller.rb]: I tried a couple of different ways on one of the previous controllers and nothing seemed to work consistently except instantiating them as empty. Happy to revisit at some point, ideally I'd like to get this working without the use of `flash` at all. It feels like it should be possible 🙂.

Comment (stuartmccoll) [reply app/controllers/project/project_non_cash_contributions_controller.rb]: Raised #137 to revisit the use of `flash`.

**Pull request merged**